File: python/sublist/sublist.py
# coding=utf-8
from __future__ import unicode_literals

__author__ = 'Cedric Zhuang'

SUBLIST, SUPERLIST, EQUAL, UNEQUAL = (
    'SUBLIST', 'SUPERLIST', 'EQUAL', 'UNEQUAL')


def check_lists(l1, l2):
    ret = UNEQUAL

    if l1 == l2:
        ret = EQUAL
    elif _is_sublist(l1, l2):
        ret = SUBLIST
    elif _is_sublist(l2, l1):
        ret = SUPERLIST
    return ret


def _is_sublist(sub, full):
    sub_len = len(sub)
    full_len = len(full)

    ret = False
    if sub_len == 0:
        ret = True
    elif sub_len < full_len:
        for fi in xrange(full_len):
            # check first char before whole for perf concern
            if full[fi] == sub[0] and full[fi:fi + sub_len] == sub:
                ret = True
                break
    return ret

File: scala/sublist/src/main/scala/Sublist.scala
import scala.annotation.tailrec

/**
  *
  * @author Cedric Zhuang
  *
  */
object Sublist extends Enumeration {
  type Rel = Value
  val Equal, Unequal, Sublist, Superlist = Value

  private def check(l1: List[Any], l2: List[Any]): Boolean = {
    @tailrec
    def loop(c1: List[Any], c2: List[Any], rest: List[Any]): Boolean =
      if (c2 == Nil) true
      else if (c1 == Nil) c2 == Nil
      else if (c1.head == c2.head) loop(c1.tail, c2.tail, rest)
      else loop(rest.tail, l2, rest.tail)

    loop(l1, l2, l1)
  }

  def sublist(l1: List[Any], l2: List[Any]): Rel =
    if (l1 == l2) Equal
    else if (l1.length > l2.length && check(l1, l2)) Superlist
    else if (l1.length < l2.length && check(l2, l1)) Sublist
    else Unequal
}
